### PR TITLE
Retry connection on ECONNRESET

### PR DESCRIPTION
**Pull request by artur-ma**
Trying to address https://github.com/fastify/help/issues/413 , still in the draft since I want to understand if I'm in the right direction, if so I will continue with the PR and will add tests

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -28,6 +28,34 @@ module.exports = fp(function from (fastify, opts, next) {
     base,
     undici: opts.undici
   })
+
+  function createRetryRequest (requestImpl, reply, retriesCount) {
+    function requestRetry (req, cb) {
+      let retries = 0
+
+      function run () {
+        requestImpl(req, function (err, res) {
```
Comment (mcollina) [inline index.js:37]: You can't retry POSTs

**Review by mcollina — CHANGES_REQUESTED** [review]

**Comment by mcollina** [comment]
(Good work anyway!)

File: index.js (hunk 1 above)
Comment (artur-ma) [reply index.js]: Retry methods restricted

**Comment by artur-ma** [comment]
I have concerns regarding `createRequestRetry` I'm creating new wrapper on each request, which may affect performance, but I'm not sure how to avoid it

**Comment by mcollina** [comment]
We would need to benchmark it.

**Comment by artur-ma** [comment]
Added tests, I will run benchmark and post results to see if there is a huge penalty 

btw, undici will not throw `ECONNRESET` it has its own error, code, should I support it too? WDYT?

**Comment by mcollina** [comment]
Yes please. I do not recall what code undici throws!

File: index.js [hunk 2]
```diff
@@ -20,14 +20,20 @@ module.exports = fp(function from (fastify, opts, next) {
     'application/json',
     ...(opts.contentTypesToEncode || [])
   ])
+
+  const retryMethods = new Set(opts.retryMethods || [
+    'GET', 'HEAD', 'PUT', 'DELETE', 'OPTIONS', 'TRACE'
```
Comment (mcollina) [inline index.js:25]: PUT and DELETE should not be retriable.

**Review by mcollina — COMMENTED** [review]
Good work! This needs docs.

**Comment by artur-ma** [comment]
without retry wrapper

image

with the wrapper

image


looks like no diff at all

File: index.js (hunk 2 above)
Comment (artur-ma) [reply index.js:25]: will remove them, although MDN says they are indepotent

**Review by mcollina — APPROVED** [review]
lgtm

File: README.md [hunk 3]
```diff
@@ -189,6 +189,13 @@ This only applies when a custom [`body`](#body) is not passed in. Defaults to:
 ]
 ```
 
+#### `retryMethods`
+
+On which methods should the connection be retried in case of socket hang up.  
+**Be aware** that setting here not idempotent method may lead to unexpected results on target.
+
+By default: `GET`, `HEAD`, `OPTIONS`, `TRACE`
```
Comment (Fdawgs) [inline README.md:197]: Can't seem to put this inside a suggestion. 😞 
I would suggest changing this to something like the following, to indicate it accepts an array of options, as at present it the documentation makes it looks like it accepts a single string:

Defaults to:
```js
[ 'GET', 'HEAD',  'OPTIONS', 'TRACE' ]
```

Comment (artur-ma) [reply README.md]: Done

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**